File: typescript/packages/reality-capture/src/service/response.ts
import { DetailedErrorResponse } from "./error";

export class Response<T> {
  status_code: number;
  error?: DetailedErrorResponse | null;
  value?: T | null;

  /**
   * A tuple containing a Service response.
   * @param status_code HTTP Status Code returned by the service.
   * @param error Optional error if the request failed.
   * @param value Optional object if the request succeeded.
   */
  constructor(status_code: number, error?: DetailedErrorResponse | null, value?: T | null) {
    this.status_code = status_code;
    this.error = error ?? null;
    this.value = value ?? null;
  }

  /**
   * Return the HTTP response status_code
   * @returns int. The HTTP response status_code
   */
  getResponseStatusCode(): number {
    return this.status_code;
  }

  /**
   * Checks whether the response is an error response.
   * @returns True if the response contains a valid error.
   */
  isError(): boolean {
    return this.error !== null && this.error !== undefined;
  }
}

File: python_sdk/src/reality_capture/service/response.py
from typing import TypeVar, Generic, Optional
from dataclasses import dataclass
from reality_capture.service.error import DetailedErrorResponse


T = TypeVar("T")

@dataclass
class Response(Generic[T]):
    """
    A tuple containing a Service response.
    """

    status_code: int
    "HTTP Status Code returned by the service."
    error: Optional[DetailedErrorResponse]
    "Optional error if the request failed."
    value: Optional[T]
    "Optional object if the request succeed."

    def get_response_status_code(self) -> int:
        """
        Return the HTTP response status_code

        Returns:
            int. The HTTP response status_code
        """
        return self.status_code

    def is_error(self) -> bool:
        """
        Checks whether the response is an error response.

        Returns:
            True if the response contains a valid error.
        """
        return self.error is not None
